10_rest/n0npax/pkg/puppy/rest_test.go

```diff
@@ -0,0 +1,226 @@
+package puppy_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	tassert "github.com/stretchr/testify/assert"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/stretchr/testify/suite"
+)
+
+type storerImpl int
+
+const (
+	syncStorer storerImpl = iota
+	memStorer
+	leveldbStorer
+)
+
+type storerSuite struct {
+	suite.Suite
+	store puppy.Storer
+	impl  storerImpl
+}
+
+func (s *storerSuite) TearDownTest() {
+	if ldbs, ok := s.store.(*store.LevelDBStore); ok {
+		ldbs.Ldb.Close()
+		os.RemoveAll(store.LevelDBPath)
+	}
+}
+
+func (s *storerSuite) SetupTest() {
+	switch s.impl {
```

Comment:
I'd use a type switch here on s.store and ensure things are properly initialised in SetupSuite.

Reply:
I may be wrong here, but:
* `s.store` is an interface. casting from an interface should be explicit.
* Moving storer initialization outside of setup test will cause sharing instance between tests. we should avoid this.

Reply:
Alternatively you pass in a `NewStorere` function that is of same type as `NewMemStore` and `NewSyncStore`
```go
type storerSuite struct {
	suite.Suite
	newStorer func() pup.Storer
        store puppy.Storer
}

func (s *storerSuite) SetupTest() {
         s.store = s.newStorer()
}

func TestStorer(t *testing.T) {
	suite.Run(t, &storerSuite{newStorer: store.NewMemStore})
	suite.Run(t, &storerSuite{newStorer: store.NewSyncStore})
}
```

Reply:
fixed, thanks for additional explanation